lib/evse_security/evse_security.cpp

```diff
@@ -880,15 +942,59 @@ GetKeyPairResult EvseSecurity::get_key_pair_internal(LeafCertificateType certifi
             return result;
         }
 
-        fs::path key_file;
-        fs::path certificate_file;
-        fs::path chain_file;
+        std::optional<X509Wrapper> latest_valid;
+        std::optional<fs::path> found_private_key_path;
+
+        // Iterate all certificates from newest to the oldest
+        leaf_certificates.for_each_chain_ordered(
+            [&](const fs::path& file, const std::vector<X509Wrapper>& chain) {
+                // Search for the first valid where we can find a key
+                if (chain.size() && chain.at(0).is_valid()) {
```

Comment:
checking for `not chain.empty()` instead of `chain.size()` might be a bit more readable